up-l1/someip.adoc

```diff
@@ -0,0 +1,221 @@
+
+When receiving a SOME/IP initiated requests:
+
+* *MUST* cache the SOME/IP request ID as well as the generated `UAttributes` for the request messages so that the response can be translated back to a SOME/IP response message
+
+
+==== Communication Status 
+
+<<commstatus-mapping>> below provides the mapping of uAttributes `commstatus` `UCode` codes to SOME/IP error codes.
+
+.Error Code Mapping
+[commstatus-mapping, width=70%]
+|===
+| UCode | SOME/IP Error Codes
+
+|`OK` |`E_OK`
+|`INVALID_ARGUMENT` |`E_WRONG_MESSAGE_TYPE` or `E_UNKNOWN_METHOD`
+|`DEADLINE_EXCEEDED` |`E_TIMEOUT`
+|`NOT_FOUND` |`E_UNKNOWN_SERVICE`
+|`UNAVAILABLE` |`E_NOT_READY`
+|`DATA_LOSS` |`E_MALFORMED_MESSAGE`
+|`INTERNAL` |`E_NOT_REACHABLE`
+|===
+
+
+=== UPayload
+
+`UPayload` hosts the application layer data that is being sent between devices. `UPayload` structure, defined in link:../basics/upayload.adoc[UPayload] contains either a copy or reference to the actual payload that is to be sent. `UPayload` also contains `UPayloadFormat` used to give a hint of the payload format (protobuf serialized, SOME/IP format, TEXT, RAW, etc...). The SOME/IP specifications however does not have an equivalent field for `UPayloadFormat` as it is assumed that the payload is serialized in the format that the other end knows how to deserialize (i.e. it is fixed per topic). As such, when converting between uProtocol and SOME/IP, the `UPayloadFormat` field *SHOULD* be ignored (left at the default of `UMESSAGE_TYPE_UNSPECIFIED`).
+
+
+=== uTransport
+
+Mapping of uTransport APIs to SOME/IP specific library APIs shall not be covered in this document given there are multiple open source libraries available for SOME/IP.
+
+
+== Application Layer Translation
+
+Application (or message payload) translation is the process of converting SOME/IP-SD subscription and discovery messages, to/from uDiscovery and uSubscription Messages. 
+
+=== uSubscription
+
+The following section will elaborate only on the translation of uSubscription messages to/from SOME/IP-SD messages. Subscription state (persistent or not) is handled in the uSubscription services and not at the transport layer or this component.
+
+The following section we will elaborate on how Eventgroup Entry types (5.1.3.1) are mapped to uSubscription messages for the subscribe
+and unsubscribe flows. 
+
+==== Common Fields
+
+<<common-field-mappings>> table below illustrates the common SOME/IP-SD EventGroup Entry fields that are present in for all SOME/IP-SD Eventgroup entry types (`SubscribeEventgroup`, `SubscribeEventGroupAck`, `SubscribeEventgroupNack`, `StopSubscribeEventGroup`). These fields are then mapped to uProtocol `UUri` attributes usined in uProtocol `UMessage` for performing subscription operations. 
+
+.Common Field Mappings
+[#common-field-mappings, width=70%]
+|===
+|Eventgroup Entry Field |UUri
+
+|Service ID |`UEntity.id`
+
+|Major Version | `UEntity.major_version`
+
+|Eventgroup ID | `UResource.id`
```

Comment:
Since Eventgroup ID and Event ID both are being mapped to UResource.id, I am guessing it means that this still doesn't support multiple event ids in an eventgroup and that eventgroup id should be equal to event id ?

Reply:
Even if we have 1 : 1 mapping for events and groups, they should be different IDs for the client library.
Everything is hardcoded in the SOME/IP service and the client library must follow those defined IDs.

I suggest we use Event ID as the primary mapping (as `UResource.id`) also for subscribe(). 
The client implementation may look up its static configuration to get additional information about Eventgroup ID to call `subcribe()`, but then events will be coming with Event ID, so subscribing to that Event ID makes more sense.

As a side effect subscribing to ECU with multiple events in the event group may trigger receive `on_message()` notifications for other events as well (those can be filtered on client level)

Reply:
I liked Ivan's suggestion.  lets keep event group logical entity (i.e. not defined in uProtocol). we can map an event group or event groups to a service.  also we can consider adding new api in sdk for registerListnergroup() which can be used for registering logical events (event id) for a services/event group.

Reply:
fixed in latest change.